File: llm-harness-js/src/stats/llm/llm-stats/cache.ts
import { mkdir, readFile, writeFile } from "node:fs/promises";
import { resolve } from "node:path";

import { type ModelStatsSelectedPayload } from "./types.js";

export const DEFAULT_OUTPUT_PATH = resolve(".cache/model_stats.json");
const CACHE_DIR = resolve(".cache");
const CACHE_TTL_SECONDS = 60 * 60 * 24;

async function writeJson(path: string, payload: unknown): Promise<void> {
  await mkdir(CACHE_DIR, { recursive: true });
  await writeFile(path, `${JSON.stringify(payload, null, 2)}\n`, "utf-8");
}

function nowEpochSeconds(): number {
  return Math.floor(Date.now() / 1000);
}

function isFreshCache(fetchedAtEpochSeconds: unknown): boolean {
  if (typeof fetchedAtEpochSeconds !== "number") {
    return false;
  }
  const ageSeconds = nowEpochSeconds() - fetchedAtEpochSeconds;
  return ageSeconds >= 0 && ageSeconds <= CACHE_TTL_SECONDS;
}

export function currentEpochSeconds(): number {
  return nowEpochSeconds();
}

export async function saveModelStatsSelectedToPath(
  payload: ModelStatsSelectedPayload,
  outputPath = DEFAULT_OUTPUT_PATH,
): Promise<void> {
  try {
    await writeJson(outputPath, payload);
  } catch {
    // Intentionally swallow cache write errors: API remains in-memory first.
  }
}

export async function loadModelStatsSelectedFromCache(
  outputPath: string,
): Promise<ModelStatsSelectedPayload | null> {
  try {
    const content = await readFile(outputPath, "utf-8");
    const payload = JSON.parse(content) as ModelStatsSelectedPayload;
    if (!Array.isArray(payload.models)) {
      return null;
    }
    if (!isFreshCache(payload.fetched_at_epoch_seconds)) {
      return null;
    }
    return payload;
  } catch {
    return null;
  }
}

File: llm_harness/stats/llm/llm_stats_stages/cache.py
"""Cache helpers for the final selected LLM stats payload."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any

from ...utils import is_fresh_epoch_seconds, now_epoch_seconds, write_json_file
from .types import ModelStatsSelectedPayload, ModelStatsSelectedPayloadModel

DEFAULT_OUTPUT_PATH = Path(".cache/llm_stats.json")
CACHE_TTL_SECONDS = 60 * 60 * 24


def current_epoch_seconds() -> int:
    return now_epoch_seconds()


def save_model_stats_selected_to_path(
    payload: ModelStatsSelectedPayload,
    output_path: Path = DEFAULT_OUTPUT_PATH,
) -> None:
    try:
        validated_payload = ModelStatsSelectedPayloadModel.model_validate(payload)
        write_json_file(output_path, validated_payload.model_dump())
    except Exception:
        return


def load_model_stats_selected_from_cache(
    output_path: Path,
) -> ModelStatsSelectedPayload | None:
    try:
        payload = json.loads(output_path.read_text(encoding="utf-8"))
        if not isinstance(payload, dict):
            return None
        models = payload.get("models")
        if not isinstance(models, list):
            return None
        fetched_at_epoch_seconds: Any = payload.get("fetched_at_epoch_seconds")
        if not is_fresh_epoch_seconds(fetched_at_epoch_seconds, CACHE_TTL_SECONDS):
            return None
        return ModelStatsSelectedPayloadModel(
            fetched_at_epoch_seconds=int(fetched_at_epoch_seconds),
            models=models,
        ).model_dump()
    except Exception:
        return None
